File: src/types/game.ts
import type { Character, CharacterNames } from './character'
import type { ItemNames } from './item'
import type {
  Subsector,
  SubsectorNames,
  Supersector,
  SupersectorNames,
} from './sectors'
import type { Ship } from './ship'

export interface Game {
  id: string
  day: number
  cycle: number
  ship: Ship
  morale: number
  maxMorale: number
  nanoshSympathy: number
  maxNanoshSympathy: number
  intel: {
    basic: number
    critical: number
  }
  characters: Map<CharacterNames, Character>
  charactersDead: Set<CharacterNames>
  shipLocation: SupersectorNames
  nanosh: {
    mainBase: SupersectorNames | null
    auxBase: Set<SupersectorNames>
    outposts: Set<SubsectorNames>
    advances: Set<SubsectorNames>
    liberationPoints: Set<SubsectorNames>
    destroyed: {
      outposts: Set<SubsectorNames>
      auxBase: Set<SupersectorNames>
    }
    assimilateEnabled: boolean
  }
  subsectors: {
    empty: Set<SubsectorNames>
  }
  sectors: Map<SupersectorNames | SubsectorNames, Supersector | Subsector>
  craftable: Set<ItemNames>
  anyMap: Map<any, any> // For anything that needs to be stored temporarily, e.g. action.bridge.command.mobilize before confirm
}

File: src/nakama/gamelogic/types/game.go
package types

type AerialUnits struct {
	Hornets int `json:"hornets"`
	Talons  int `json:"talons"`
}

type Intel struct {
	Basic    int `json:"basic"`
	Critical int `json:"critical"`
}

type NanoshDestroyed struct {
	Outposts map[SubsectorName]bool   `json:"outposts"`
	AuxBase  map[SupersectorName]bool `json:"auxBase"`
}

type Nanosh struct {
	MainBase          *SupersectorName         `json:"mainBase"`
	AuxBase           map[SupersectorName]bool `json:"auxBase"`
	Outposts          map[SubsectorName]bool   `json:"outposts"`
	Advances          map[SubsectorName]bool   `json:"advances"`
	LiberationPoints  map[SubsectorName]bool   `json:"liberationPoints"`
	Destroyed         NanoshDestroyed          `json:"destroyed"`
	AssimilateEnabled bool                     `json:"assimilateEnabled"`
	AerialUnits       AerialUnits              `json:"aerialUnits"`
}

type SubsectorsInfo struct {
	Empty map[SubsectorName]bool `json:"empty"`
}

type Game struct {
	ID                string                       `json:"id"`
	Day               int                          `json:"day"`
	Cycle             int                          `json:"cycle"`
	Ship              Ship                         `json:"ship"`
	Morale            int                          `json:"morale"`
	MaxMorale         int                          `json:"maxMorale"`
	NanoshSympathy    int                          `json:"nanoshSympathy"`
	MaxNanoshSympathy int                          `json:"maxNanoshSympathy"`
	Intel             Intel                        `json:"intel"`
	Characters        map[CharacterName]*Character `json:"characters"`
	CharactersDead    map[CharacterName]bool       `json:"charactersDead"`
	ShipLocation      SupersectorName              `json:"shipLocation"`
	Nanosh            Nanosh                       `json:"nanosh"`
	Subsectors        SubsectorsInfo               `json:"subsectors"`
	Sectors           map[string]any               `json:"sectors"` // Can be Supersector or Subsector
	Craftable         map[ItemName]bool            `json:"craftable"`
	Expedition        *Expedition                  `json:"expedition"`
	BattleState       *BattleState                 `json:"battleState"`
	AnyMap            map[string]any               `json:"anyMap"` // For temporary storage
}
